Add support for files in OPFS

ansemjo opened the pull request:
This could partly fix #32 by allowing to use [`FileSystemSyncAccessHandle`s](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle) in Workers for single files opened from [OPFS](https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system) as [noted in my last comment](https://github.com/bjorn3/browser_wasi_shim/issues/32#issuecomment-1612637810). Usage example:

```ts
// open synchronous handle to a file in opfs
const opfs = await navigator.storage.getDirectory();
const file = await opfs.getFileHandle("example.txt");
const handle = await file.createSyncAccessHandle();

// populate the virtual filesystem for shim
const fds: Array<Fd> = [
  new OpenFile(new File([])), // stdin
  new OpenFile(new File([])), // stdout
  new OpenFile(new File([])), // stderr
  new PreopenDirectory("/", {
    "example.txt": new OPFSFile(handle),
  }),
];
const shim = new WASI(args, envs, fds);
```

~~**This should currently be considered *very* WIP, as I've developed this out-of-tree in my project and then tried to copy it into this fork. That means I haven't actually run the code here yet.**~~ In my project I could successfully read a small file though: 
![image](https://github.com/bjorn3/browser_wasi_shim/assets/11139925/978477d4-5c1d-4760-b9c8-abf354eefcef)

ansemjo commented:
* One problem that I already noticed, is that the file is never actually closed though. My experiment in Rust just uses `std::fs::read()` to read the contents and so apparently does not explicitly call `fd_close()` anywhere. The result is that trying to access the file a second time fails with a `NoModificationAllowedError: No modification allowed`.
* Also, I should probably implement more of the `Fd` methods ... `fd_sync()` comes to mind, as the `FileSystemSyncAccessHandle` brings a `flush()` method. We should probably use that.

bjorn3 commented:
Thanks a lot for working on this!

> One problem that I already noticed, is that the file is never actually closed though. My experiment in Rust just uses std::fs::read() to read the contents and so apparently does not explicitly call fd_close() anywhere. The result is that trying to access the file a second time fails with a NoModificationAllowedError: No modification allowed.

Weird that it is never closed. Can you use the strace wrapper to show all syscalls that are made? In any case there should be code added to handle opening the same file multiple times. At least once directories are supported.

ansemjo commented on src/fs_core.ts:
```diff
@@ -21,19 +21,46 @@ export class File {
+  write(buffer: ArrayBuffer | ArrayBufferView, options?: { at: number }): number;
+}
+
+export class OPFSFile {
```
Yep, you're right! I'm currently trying to fix the exported typings, after figuring out how to `yarn link` my local fork into my project ...

ansemjo commented:
> Thanks a lot for working on this!
> 
> > One problem that I already noticed, is that the file is never actually closed though. My experiment in Rust just uses std::fs::read() to read the contents and so apparently does not explicitly call fd_close() anywhere. The result is that trying to access the file a second time fails with a NoModificationAllowedError: No modification allowed.
> 
> Weird that it is never closed. Can you use the strace wrapper to show all syscalls that are made? In any case there should be code added to handle opening the same file multiple times. At least once directories are supported.

Maybe I should clarify: by opening a second time I mean starting a second WebAssembly instance, which also tries to open the file from OPFS. So technically, the previous instance simply still holds the lock on the file. My expectation would have been that at the end of `wasi.start(instance)` all opened files would be closed again but I am not sure if that'd be correct behaviour.

Anyway, how do I use the strace wrapper? Just wrap the `"wasi_snapshot_preview1"` import as seen in the example?
```
let inst = await WebAssembly.instantiate(wasm, {
  "wasi_snapshot_preview1": strace(w.wasiImport, ["fd_prestat_get"]),
});
```

bjorn3 commented:
> Maybe I should clarify: by opening a second time I mean starting a second WebAssembly instance, which also tries to open the file from OPFS. So technically, the previous instance simply still holds the lock on the file. My expectation would have been that at the end of wasi.start(instance) all opened files would be closed again but I am not sure if that'd be correct behaviour.

For reactor style wasi modules as handled by `.initialize()` fds should be kept. For command style wasi modules as handled by `.start()` fds could be closed at the end. That would be a breaking change though. Maybe you could add a FIXME to `start()` to do this in v0.3?

> Anyway, how do I use the strace wrapper? Just wrap the "wasi_snapshot_preview1" import as seen in the example?

Yes: https://github.com/bjorn3/browser_wasi_shim/blob/ee4f28732cd86bbebeb92969af7cee6d30ac5225/examples/rustc.html#L129-L131

ansemjo commented:
Aha! Mea culpa .. the Rust experiment actually *does* close the file as seen here:

![image](https://github.com/bjorn3/browser_wasi_shim/assets/11139925/c0a88028-19f6-4ee5-b1ac-98f9e504820c)

This is the code that is executed there:

```rust
fn main() {
  println!("This is Rust.");

  // print commandline arguments
  let args: Vec<String> = std::env::args().collect();
  let name = args[0].clone();
  println!("file '{name}' was called with arguments: {args:?}");

  // print environment variables
  println!("environment variables:");
  for (key, value) in std::env::vars() {
    println!(" - {}: {}", key, value);
  };

  // list root filesystem
  match std::fs::read_dir("/") {
    Err(e) => println!("ERR: couldn't open directory '/': {e}"),
    Ok(iter) => {
      println!("listing '/' contents:");
      for item in iter {
        println!(" - {}", item.unwrap().path().display());
      }
    }
  };

  // try to read a specific file
  const FILENAME: &str = "/hello.txt";
  match std::fs::read(FILENAME) {
    Err(e) => println!("ERR: couldn't open file '{FILENAME}': {e}"),
    Ok(content) => {
      let text = String::from_utf8_lossy(&content);
      println!("'{FILENAME}': {text}");
    },
  }

}
```

My TinyGo experiment however, does not ..

![image](https://github.com/bjorn3/browser_wasi_shim/assets/11139925/2deb8b4a-ab89-42aa-b658-a4dc034d3cb0)

```go
package main

import (
	"fmt"
	"io"
	"os"
)

func main() {
	fmt.Println("This is TinyGo.")

	// print commandline arguments
	args := os.Args
	name := args[0]
	fmt.Printf("file '%s' was called with arguments: %v\n", name, args)

	// print environment variables
	fmt.Println("environment variables:")
	for _, env := range os.Environ() {
		fmt.Printf(" - %s\n", env)
	}

	// list root filesystem
	contents, err := os.ReadDir("/")
	if err != nil {
		fmt.Printf("ERR: couldn't open directory '/': %s\n", err)
	} else {
		fmt.Println("listing '/' contents:")
		for _, item := range contents {
			fmt.Printf(" - %s\n", item.Name())
		}
	}

	// try to read a specific file
	const FILENAME = "/hello.txt"
	file, err := os.Open(FILENAME)
	if err != nil {
		fmt.Printf("ERR: couldn't open file '%s': %s\n", FILENAME, err)
	} else {
		bytes, err := io.ReadAll(file)
		if err != nil {
			fmt.Printf("ERR: couldn't read file '%s': %s\n", FILENAME, err)
		} else {
			fmt.Printf("'%s': %s\n", FILENAME, string(bytes))
		}
	}

}
```

I mean, yeah .. I neither call nor defer `file.Close()` anywhere. I've added a FIXME in the last commit.

ansemjo commented on src/fs_fd.ts:
```diff
@@ -178,6 +262,8 @@ export class OpenDirectory extends Fd {
       // @ts-ignore
       return { ret: 0, fd_obj: new OpenFile(entry) };
+    } else if (entry instanceof OPFSFile) {
+      return { ret: 0, fd_obj: new OpenOPFSFile(entry) }
```
Don't they all already derive from `Fd`? We could just use duck-typing to check if the `entry` has an `open()` property and call that.

ansemjo replied:
Well, none of `File | SyncOPFSFile | Directory` derived from `Fd` but Typescript handles the type union to check if `.open()` exists just fine without a common base class: 512a46d

ansemjo commented:
It is definitely incorrect to close the handle in `fd_close()` though, as that makes the file unusable for the rest of the execution. So this should also be handled in `SyncOPFSFile` after `WASI.start()`.

ansemjo commented on src/fs_core.ts:
```diff
@@ -37,10 +39,13 @@ export interface FileSystemSyncAccessHandle {
 
-  open() {
-    return new OpenSyncOPFSFile(this);
+  open(fd_flags: number) {
```
@bjorn3 Added handling for `FDFLAGS_APPEND` here and in `File` to be able to correctly *append* to existing files.

ansemjo commented:
I've tested writes and caught an OOB bug in the process; added support for `FDFLAGS_APPEND`; and removed my `console.log` lines.

Regarding the closing of the `FileSystemSyncAccessHandle`: I'm thinking that maybe we just shouldn't bother with it, since the way I have implemented it here, I already accept a handle. Which means it was created externally and maybe we *shouldn't* make assumptions and close it.

The `Inode` base class idea I'd leave for another pull request. 😛 

If you don't have any further comments, I think you might merge this now as the basic read / write / seek / append operations work now, as far as I can tell.

bjorn3 commented on src/fs_core.ts:
```diff
@@ -21,19 +28,57 @@ export class File {
   }
 }
 
+// shim https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle
```
```suggestion
// Shim for https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle
// This is not part of the public interface.
```

Also looks like updating to typescript 5 is enough to make this available: https://github.com/microsoft/TypeScript-DOM-lib-generator/issues/1471 That should probably be a separate PR.

ansemjo replied:
Huh. Maybe there's something else you need to set because this is only available in WebWorkers? My project is on TypeScript 5.0.4 and doesn't seem to include this type. I've already used [`@types/web`](https://www.npmjs.com/package/@types/web) to get support for the WebTransport API but the editor still complains about `FileSystemSyncAccessHandle`.

ansemjo replied:
Indeed, [you need to tell Typescript](https://joshuatz.com/posts/2021/strongly-typed-service-workers/) the the file should use the `webworker` lib:

```ts
/// <reference no-default-lib="true"/>
/// <reference lib="es2020" />
/// <reference lib="webworker" />

...
```

And as soon as I did that it started complaining that the types are incompatible 🙃 
![image](https://github.com/bjorn3/browser_wasi_shim/assets/11139925/9bbe4292-665c-4167-8f33-da4fcfe0ed17)

But then adding `/// <reference lib="webworker" />` at the top of `fd_core.ts` and installing `typescript@^5.0.4` fixes that for me. I'm not adding that as a commit though because you said that'd be for a separate PR.

I've just fixed the "shim" for now because while MDN says these return a `Promise<undefined>`, which is obviously an error, I thought they meant just `undefined`. The specification and typings seem to use `void` though.

bjorn3 commented on src/index.ts:
```diff
@@ -2,6 +2,6 @@ import WASI from "./wasi.js";
 export { Fd } from './fd.js';
-export { File, Directory } from "./fs_core.js";
-export { OpenFile, OpenDirectory, PreopenDirectory } from "./fs_fd.js";
+export { File, SyncOPFSFile as OPFSFile, Directory } from "./fs_core.js";
```
```suggestion
export { File, SyncOPFSFile, Directory } from "./fs_core.js";
```

bjorn3 commented:
I'm going to test this myself tomorrow. If I don't see anything weird I will merge it.

bjorn3 commented on src/fs_core.ts:
```diff
@@ -21,19 +28,58 @@ export class File {
+  write(buffer: ArrayBuffer | ArrayBufferView, options?: { at: number }): number;
+}
+
+export class SyncOPFSFile {
```
```suggestion
// Synchronous access to an individual file in the origin private file system.
// Only allowed inside a WebWorker.
export class SyncOPFSFile {
```

bjorn3 commented:
Do you have more changes you want or shall I publish a new release with this?

ansemjo commented:
> Got it working locally.

Great! 😊 

> Do you have more changes you want or shall I publish a new release with this?

Right now I'm just glad you found this useful enough to merge so quickly. :) I see you reopened the issue to track work on the async API / whole directories .. that'd be *great* but I don't think I can justify the effort to even attempt that right now.

It probably should've been part of this PR but maybe add another usage example on how to use this new `SyncOPFSFile`, since the only way to find it right now is to look through the code or find this issue. Is your test code suitable for this?

Other than that I don't have anything that should be a blocker for a new release, I think.

bjorn3 commented:
> I see you reopened the issue to track work on the async API / whole directories .. that'd be great but I don't think I can justify the effort to even attempt that right now.

No problem.

> It probably should've been part of this PR but maybe add another usage example on how to use this new SyncOPFSFile, since the only way to find it right now is to look through the code or find this issue. Is your test code suitable for this?

The code I used to test it was using the OPFS file as stderr for rustc.wasm and then manually verifying that it wrote to the file in the browser console. Not really something usable as test for others.
